### PR TITLE
Add JWT decoding support class

### PR DESCRIPTION
**Pull request by trevrosen**
## Concepts to Understand
* [JWT](http://jwt.io)
* [HMAC](https://en.wikipedia.org/wiki/Hash-based_message_authentication_code)

## What
Support for checking signatures and decoding Clortho-provided JWT claims data

## Why
When we go live with [Clortho](https://github.com/bypasslane/clortho), we will be providing JWT-based tokens to all our authenticated clients, and will be slowly phasing out the practice of hitting the auth service for every request in favor of enabling the token itself to carry around its own notion of validity.

## Why not use a JWT library?
Signed JWTs are simply base64-encoded JSON with a signature block appended to the end. External libraries require support of much more of the spec than we are using and can't make assumptions about what can be hardcoded. In short, they represent bloat, and as you can see from this class, JWTs just aren't that hard to deal with.

### CODE REVIEW

File: spec/fixtures/jwt/user_token_claims.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (jspies) [inline spec/fixtures/jwt/user_token_claims.json]: Is this the def for session fields inclusion? (I know it's a fixture, but I don't see it anywhere else)

File: lib/bypass_auth/jwt.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  def claims_data
+    if instance_variable_get(:@claims_data).nil?
+      if signature_valid?
+        @claims_data = JSON.parse(Base64.urlsafe_decode64(claims_string), symbolize_names: true)
+      end
+    end
+    @claims_data
+  end
+
+  # The key used for generating the HMAC signature of the token must be shared by both Clortho
```
Comment (jspies) [inline lib/bypass_auth/jwt.rb]: let's privatize what we can so the interface is simple

File: spec/fixtures/jwt/user_token_claims.json (hunk 1 above)
Comment (trevrosen) [inline spec/fixtures/jwt/user_token_claims.json]: There is not in the client, but there is in Clortho. The client only cares about verify and decoding, but doesn't enforce/validate contents.

Comment (jspies) [inline spec/fixtures/jwt/user_token_claims.json]: gotcha

**Comment by jspies** [comment]
🚢

**Pull request merged**